### PR TITLE
Improve JSON schema draft 4 support

### PR DESCRIPTION
**Pull request by krismolendyke**
Thanks for creating a Python JSON schema validation library that isn't slow!

Aside from the good performance, and the differences noted in the documentation, I wanted to know exactly how well this library validated against the official JSON schema (draft 4) test suite.

To do that I have made the following changes:

- Add https://github.com/json-schema-org/JSON-Schema-Test-Suite submodule
- Add test suite runner script
- Add test suite runner configuration specifying a subset of the draft 4 test suite
- Update code generator to pass tests specified in configuration

To support those changes:

- Add `extras_require` to `setup.py` with `test` requirements
- Update `Makefile` to use `pip` to install `test` requirements
- Fix `performance.py` script to properly `dedent` `code` for `timeit` run

To nitpick ;p :

- Fix a typographical error, `spcified` → `specified`

All current `make test` target tests still pass via `pytest`.

This notably skips these draft 4 tests via configuration:

```
definitions.json
dependencies.json
optional/bignum.json
optional/ecmascript-regex.json
optional/format.json
optional/zeroTerminatedFloats.json
ref.json
refRemote.json
uniqueItems.json
```

`uniqueItems.json` partially passes but fails on nested objects.  Passing that test will come at cost to performance.

Example test suite execution:

```sh
$ ./json_schema_test_suite.py json-schema-test-suite-draft-4.conf
✔ additionalItems.json
✔ additionalProperties.json
✔ allOf.json
✔ anyOf.json
⛔ bignum.json
✔ default.json
⛔ definitions.json
⛔ dependencies.json
⛔ ecmascript-regex.json
✔ enum.json
⛔ format.json
✔ items.json
✔ maxItems.json
✔ maxLength.json
✔ maxProperties.json
✔ maximum.json
✔ minItems.json
✔ minLength.json
✔ minProperties.json
✔ minimum.json
✔ multipleOf.json
✔ not.json
✔ oneOf.json
✔ pattern.json
✔ patternProperties.json
✔ properties.json
⛔ ref.json
⛔ refRemote.json
✔ required.json
✔ type.json
⛔ uniqueItems.json
⛔ zeroTerminatedFloats.json

Schema exceptions:

⛔ ref.json: expected an indented block (<string>, line 29): 'return data'
⛔ refRemote.json: expected an indented block (<string>, line 11): 'return data'

Summary of 358 tests:

Failures:

✘ FALSE_POSITIVE    0   0.0%
✘ FALSE_NEGATIVE    0   0.0%
⚠ UNDEFINED         0   0.0%
                    0   0.0%

Passes:

✔ TRUE_POSITIVE   125  52.7%
✔ TRUE_NEGATIVE   112  47.3%
                  237 100.0%

⛔ Ignored:        121
Coverage:     237/358  66.2%
```

I may have survived...

```
#    ___
#    \./     DANGER: This module implements some code generation
# .--.O.--.          techniques involving string concatenation.
#  \/   \/           If you look at it, you might die.
#
```

### CODE REVIEW

File: fastjsonschema/generator.py [hunk 1]
```diff
@@ -271,29 +274,43 @@ def generate_not(self):
...
         self.create_variable_with_length()
         with self.l('if {variable}_len > {maxLength}:'):
             self.l('raise JsonSchemaException("{name} must be shorter than or equal to {maxLength} characters")')
 
     def generate_pattern(self):
+        with self.l('if not isinstance({variable}, str):'):
+            self.l('return {variable}')
         self._compile_regexps['{}_re'.format(self._variable)] = re.compile(self._definition['pattern'])
-        with self.l('if not {variable}_re.match({variable}):'):
+        with self.l('if not {variable}_re.search({variable}):'):
```
Comment (horejsek) [inline fastjsonschema/generator.py]: Why this change? :-)

File: fastjsonschema/generator.py [hunk 2]
```diff
@@ -367,29 +397,40 @@ def generate_items(self):
...
+            self.l('return {variable}')
         self.create_variable_with_length()
         with self.l('if not all(prop in {variable} for prop in {required}):'):
             self.l('raise JsonSchemaException("{name} must contain {required} properties")')
 
     def generate_properties(self):
+        with self.l('try:'):
+            self.l('{variable}.keys()')
+        with self.l('except AttributeError:'):
+            self.l('return {variable}')
```
Comment (horejsek) [inline fastjsonschema/generator.py]: Wouldn't be simple condition faster than try-except block?

File: fastjsonschema/generator.py [hunk 3]
```diff
@@ -399,13 +440,30 @@ def generate_properties(self):
...
+                {'patternProperties': self._definition['patternProperties']},
+                self._variable,
+                '{}.{}'.format(self._variable_name, 'patternProperties'),
+            )
+            self.l('pattern_keys = set()')
+            with self.l('for key in {variable}_keys:'):
+                for pattern in self._definition['patternProperties'].keys():
+                    with self.l('if globals()["{}_re"].search(key):', pattern):
+                        self.l('pattern_keys.add(key)')
+            self.l('{variable}_keys -= pattern_keys')
```
Comment (horejsek) [inline fastjsonschema/generator.py]: I don't understand yet why this block is here, why methods later are not enough.

File: fastjsonschema/generator.py (hunk 1 above)
Comment (krismolendyke) [reply fastjsonschema/generator.py]: > When defining the regular expressions, it’s important to note that the string is considered valid if the expression matches anywhere within the string.

– https://spacetelescope.github.io/understanding-json-schema/reference/string.html#index-2

> Python offers two different primitive operations based on regular expressions: re.match() checks for a match only at the beginning of the string, while re.search() checks for a match anywhere in the string

– http://devdocs.io/python~3.6/library/re#search-vs-match

File: fastjsonschema/generator.py (hunk 2 above)
Comment (krismolendyke) [reply fastjsonschema/generator.py]: That's likely.  I will refactor any more of these blocks to `isinstance` checks to align with the rest of the code.  My mistake.

Comment (horejsek) [reply fastjsonschema/generator.py]: No worry, great, thanks!

File: fastjsonschema/generator.py (hunk 1 above)
Comment (horejsek) [reply fastjsonschema/generator.py]: I see! I misinterpreted the validity of regexp in JSON schema. Thanks for the fix!

File: fastjsonschema/generator.py (hunk 3 above)
Comment (horejsek) [reply fastjsonschema/generator.py]: :-) Yep, those all properties are funny one. Will look at it in more detail on Monday.

File: fastjsonschema/generator.py [hunk 4]
```diff
@@ -414,3 +470,46 @@ def generate_properties(self):
                         '{}_value'.format(self._variable),
                         '{}.{{{}_key}}'.format(self._variable_name, self._variable),
                     )
+
+    def generate_additional_properties(self):
+        with self.l('if not isinstance({variable}, dict):'):
+            self.l('return {variable}')
```
Comment (horejsek) [inline fastjsonschema/generator.py]: This is done already by object type on the beginning. No need to do it again.

File: fastjsonschema/generator.py [hunk 5]
```diff
@@ -414,3 +470,46 @@ def generate_properties(self):
...
+                        add_prop_definition,
+                        '{}_value'.format(self._variable),
+                        '{}.{{{}_key}}'.format(self._variable_name, self._variable),
+                    )
+            if 'default' in add_prop_definition:
+                self.l('else: {variable}["{}"] = {}', key, repr(add_prop_definition['default']))
+
+    def generate_pattern_properties(self):
+        with self.l('if not isinstance({variable}, dict):'):
+            self.l('return {variable}')
```
Comment (horejsek) [inline fastjsonschema/generator.py]: This is created already by object type on the beginning. No need to do it again.

File: fastjsonschema/generator.py [hunk 6]
```diff
@@ -414,3 +470,46 @@ def generate_properties(self):
...
+            with self.l('for key in {variable}_keys:'):
+                for pattern in self._definition['patternProperties'].keys():
+                    with self.l('if globals()["{}_re"].search(key):', pattern):
+                        self.l('pattern_keys.add(key)')
+            self.l('{variable}_keys -= pattern_keys')
+
+        add_prop_definition = self._definition["additionalProperties"]
+        if add_prop_definition:
+            with self.l('for {variable}_key in {variable}_keys:'):
+                with self.l('if {variable}_key not in "{}":', self._definition.get('properties', [])):
```
Comment (horejsek) [inline fastjsonschema/generator.py]: This generates wrong code:

        if data_key not in "{'foo': {'type': 'array', 'maxItems': 3}, 'bar': {'type': 'array'}}":

I think this condition shouldn't even be here as in `{variable}_keys` is only the additional properties already.

**Comment by horejsek** [comment]
I checked the code in details. There is many issues with `patternProps` and `additionalProps`. For example I looked how is code generated for this JSON schema:

    {'properties': {'foo': {'type': 'array', 'maxItems': 3}, 'bar': {'type': 'array'}}, 'patternProperties': {'f.o': {'minItems': 2}}, 'additionalProperties': {'type': 'integer'}}

Solution you provided:

```
def func(data):
    NoneType = type(None)
    if not isinstance(data, dict):
        return data
    data_keys = set(data.keys())
    if "foo" in data_keys:
        data_keys.remove("foo")
        data_foo = data["foo"]
        if not isinstance(data_foo, (list)):
            raise JsonSchemaException("data.foo must be array")
        if not isinstance(data_foo, list):
            return data_foo
        data_foo_len = len(data_foo)
        if data_foo_len > 3:
            raise JsonSchemaException("data.foo must contain less than or equal to 3 items")
    if "bar" in data_keys:
        data_keys.remove("bar")
        data_bar = data["bar"]
        if not isinstance(data_bar, (list)):
            raise JsonSchemaException("data.bar must be array")
    if not isinstance(data, dict):
        return data

    for key, val in data.items():
        if globals()["f.o_re"].search(key):
            if not isinstance(val, list):
                return val
            val_len = len(val)
            if val_len < 2:
                raise JsonSchemaException(""+"data.patternProperties.{key}".format(**locals())+" must contain at least 2 items")
    pattern_keys = set()
    for key in data_keys:
        if globals()["f.o_re"].search(key):
            pattern_keys.add(key)
    data_keys -= pattern_keys
    for data_key in data_keys:
        data_value = data.get(data_key)
        if not isinstance(data_value, (int)) or isinstance(data_value, bool):
            raise JsonSchemaException(""+"data.{data_key}".format(**locals())+" must be integer")
    if not isinstance(data, dict):
        return data
    
    for key, val in data.items():
        if globals()["f.o_re"].search(key):
            if not isinstance(val, list):
                return val
                
            if val_len < 2:
                raise JsonSchemaException(""+"data.{key}".format(**locals())+" must contain at least 2 items")
    if not isinstance(data, dict):
        return data
    data_keys = set(data.keys())
    pattern_keys = set()
    for key in data_keys:
        if globals()["f.o_re"].search(key):
            pattern_keys.add(key)
    data_keys -= pattern_keys
    for data_key in data_keys:
        if data_key not in "{'foo': {'type': 'array', 'maxItems': 3}, 'bar': {'type': 'array'}}":
            data_value = data.get(data_key)
            if not isinstance(data_value, (int)) or isinstance(data_value, bool):
                raise JsonSchemaException(""+"data.{data_key}".format(**locals())+" must be integer")
    return data
```

Code which should be correct if I didn't make any mistake but looks good:

```
def func(data):
    NoneType = type(None)
    if not isinstance(data, dict):
        return data
    data_keys = set(data.keys())
    if "foo" in data_keys:
        data_keys.remove("foo")
        data_foo = data["foo"]
        if not isinstance(data_foo, (list)):
            raise JsonSchemaException("data.foo must be array")
        if not isinstance(data_foo, list):
            return data_foo
        data_foo_len = len(data_foo)
        if data_foo_len > 3:
            raise JsonSchemaException("data.foo must contain less than or equal to 3 items")
    if "bar" in data_keys:
        data_keys.remove("bar")
        data_bar = data["bar"]
        if not isinstance(data_bar, (list)):
            raise JsonSchemaException("data.bar must be array")
    for key, val in data.items():
        if globals()["f.o_re"].search(key):
            if key in data_keys:
                data_keys.remove(key)
            if not isinstance(val, list):
                return val
            val_len = len(val)
            if val_len < 2:
                raise JsonSchemaException(""+"data.{key}".format(**locals())+" must contain at least 2 items")
    for data_key in data_keys:
        data_value = data.get(data_key)
        if not isinstance(data_value, (int)) or isinstance(data_value, bool):
            raise JsonSchemaException(""+"data.{data_key}".format(**locals())+" must be integer")
    return data
```

I commented specific problems in the diff. Do you want to fix it or should I?

**Comment by krismolendyke** [comment]
Thanks for the review! I'm not working today, I'll have a closer look
tomorrow. If you want to make the changes and the test suites still pass
that would be great!

**Comment by horejsek** [comment]
Ok, no problem. I will leave it to you as I know it's good feeling to finish work. :-) Anyway I can help if needed by end of the week.

File: fastjsonschema/generator.py [hunk 7]
```diff
@@ -414,3 +470,46 @@ def generate_properties(self):
...
+                        '{}.{{{}_key}}'.format(self._variable_name, self._variable),
+                    )
+            if 'default' in add_prop_definition:
+                self.l('else: {variable}["{}"] = {}', key, repr(add_prop_definition['default']))
+
+    def generate_pattern_properties(self):
+        with self.l('if not isinstance({variable}, dict):'):
+            self.l('return {variable}')
+        for pattern, definition in self._definition['patternProperties'].items():
+            self._compile_regexps['{}_re'.format(pattern)] = re.compile(pattern)
```
[the remark this answers is not in the document]
Comment (krismolendyke) [reply fastjsonschema/generator.py]: I tried the non-`globals()` index-based approach first and I worried about ordering during key iteration so I refactored to this may-based approach.  Ordering shouldn't be an issue if the `dict` is never mutated but I didn't know how to be sure about that beyond converting the schema definition to an `OrderedDict` and that felt like too much change.

I don't love polluting `globals` but this seemed like the lesser of evils.  A local solution would be faster but I haven't profiled the difference.

File: fastjsonschema/generator.py [hunk 8]
```diff
@@ -414,3 +470,46 @@ def generate_properties(self):
...
+            self.l('return {variable}')
+        self.l('{variable}_keys = set({variable}.keys())')
+
+        if 'patternProperties' in self._definition:
+            self.l('pattern_keys = set()')
+            with self.l('for key in {variable}_keys:'):
+                for pattern in self._definition['patternProperties'].keys():
+                    with self.l('if globals()["{}_re"].search(key):', pattern):
+                        self.l('pattern_keys.add(key)')
+            self.l('{variable}_keys -= pattern_keys')
```
[the remark this answers is not in the document]
Comment (krismolendyke) [reply fastjsonschema/generator.py]: You're correct.  I may have implemented these "out of order" based on how the JSON schema test suite executes.  We should be able to simply `return` here.  Good catch.

File: fastjsonschema/generator.py [hunk 9]
```diff
@@ -414,3 +470,46 @@ def generate_properties(self):
                         '{}_value'.format(self._variable),
                         '{}.{{{}_key}}'.format(self._variable_name, self._variable),
                     )
+
+    def generate_additional_properties(self):
+        with self.l('if not isinstance({variable}, dict):'):
+            self.l('return {variable}')
+        self.l('{variable}_keys = set({variable}.keys())')
```
[the remark this answers is not in the document]
Comment (krismolendyke) [reply fastjsonschema/generator.py]: I understand your comments, but I think this is required due to this test:

```sh
$ ./json_schema_test_suite.py JSON-Schema-Test-Suite/tests/draft4/additionalProperties.json 
✔ additionalProperties.json
  ⚠ additionalProperties can exist by itself
    ⚠ UNDEFINED NameError an additional valid property is valid: name 'data_keys' is not defined
    ⚠ UNDEFINED NameError an additional invalid property is invalid: name 'data_keys' is not defined

Summary of 12 tests:

Failures:

✘ FALSE_POSITIVE    0   0.0%
✘ FALSE_NEGATIVE    0   0.0%
⚠ UNDEFINED         2  16.7%
                    2  16.7%

Passes:

✔ TRUE_POSITIVE     8  66.7%
✔ TRUE_NEGATIVE     2  16.7%
                   10  83.3%

⛔ Ignored:          0
Coverage:      12/12 100.0%
```

Which is a pretty odd case, but is apparently valid according to the spec.

File: fastjsonschema/generator.py (hunk 6 above)
Comment (krismolendyke) [reply fastjsonschema/generator.py]: You are right, I meant to just check if `{variable}_key` was not in the `properties` keys.  I'll fix that.

I do think we need this check though for this test case:

```
$ ./json_schema_test_suite.py JSON-Schema-Test-Suite/tests/draft4/additionalProperties.json 
✘ additionalProperties.json
  ✘ additionalProperties allows a schema which should validate
    ✘ FALSE_NEGATIVE JsonSchemaException no additional properties is valid: data.foo must be boolean
    ✘ FALSE_NEGATIVE JsonSchemaException an additional valid property is valid: data.bar must be boolean

Summary of 2 tests:

Failures:

✘ FALSE_POSITIVE    0   0.0%
✘ FALSE_NEGATIVE    2 100.0%
⚠ UNDEFINED         0   0.0%
                    2 100.0%

Passes:

✔ TRUE_POSITIVE     0   0.0%
✔ TRUE_NEGATIVE     0   0.0%
                    0   0.0%

⛔ Ignored:          0
Coverage:       2/2 100.0%
```

File: fastjsonschema/generator.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (horejsek) [reply fastjsonschema/generator.py]: This should be connected to previous comment where keys should be managed once.

**Comment by krismolendyke** [comment]
Thanks again for all the help and review.  I pushed a couple fixes and responded to several review comments.  I'm happy to make additional changes if necessary.

At this point, given how the draft 4 spec tests are written, and how confusing the interaction between `properties`, `patternProperties` and `additionalProperties` is, I think we should failing test cases before changing too much more.  The JSON schema test suite is really all I have to go to determine "correctness."  What are your thoughts?

**Comment by horejsek** [comment]
There are still some issue. I merged your changes and continue work on it.

**Comment by krismolendyke** [comment]
Excellent.  I'm happy to help out.  Thanks again!

P.S. maybe Travis or similar CI would make future pull requests easier to review?

**Comment by horejsek** [comment]
I just pushed new code to master. I fixed several bugs. One of them was that any early return would mess up the data in deep conditions. Undeclared variable. I simplified pattern and additional properties. Added support of formats and base support of refs. Also moved tests to pytest so it's easier to work with. It still need to be more tested, I know about some unit tests which has to be made and also I want to test if there is still some performance issue.

**Comment by horejsek** [comment]
I will not have time for it next two weeks. If you want, you can set up travis, do unit tests for early return and came up with more other tests for edge cases. Also you can check the performance of changes if it still the same as before.

**Comment by horejsek** [comment]
I released version 1.2.
